src/datepicker/index.es

```diff
@@ -0,0 +1,189 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+
+import { screenPosition, screenSize, size } from 'nebenan-helpers/lib/dom';
+import keymanager from 'nebenan-helpers/lib/keymanager';
+import eventproxy from 'nebenan-helpers/lib/eventproxy';
+
+import Picker from 'nebenan-react-datepicker/lib';
+import { bindTo } from '../utils';
+
+import InputComponent from '../base';
+import theme from './theme';
+import { getValue } from './utils';
+
+class Datepicker extends InputComponent {
+  constructor(props) {
+    super(props);
+    bindTo(this,
+      'handleGlobalClick',
+      'deactivate',
+      'handleSelect',
+      'handleClick',
+      'handleClear',
+    );
+  }
+
+  componentWillUnmount() {
+    this.deactivate();
+    super.componentWillUnmount();
+  }
+
+  getDefaultState(props) {
+    return {
+      ...super.getDefaultState(props),
+      isVisible: false,
+      isTop: false,
+    };
+  }
+
+  handleGlobalClick(event) {
+    if (!this.isComponentMounted) return;
+    const isOutside = !this.els.container.contains(event.target);
+    if (isOutside) this.deactivate();
+  }
+
+  getPosition() {
+    const { container } = this.els;
+
+    const { top } = screenPosition(container);
+    const { height } = size(container);
+    const { height: screenHeight } = screenSize(global);
+
+    const availableSpace = screenHeight - top;
+    const neededSpace = height * 2;
+    const isTop = neededSpace > availableSpace;
+
+    return { isTop };
+  }
+
+  activate() {
+    if (this.isVisible()) return;
+
+    this.stopListeningToKeys = keymanager('esc', this.deactivate);
+    this.stopListeningToClicks = eventproxy('click', this.handleGlobalClick);
+
+    const updater = () => {
+      const { isTop } = this.getPosition();
+      return { isVisible: true, isTop };
+    };
+
+    this.setState(updater);
```

Comment:
this state update doesn't depend on current state, i think you can call it synchronously

Reply:
I used a updater function because it depends on the current position in dom. Wouldn't want to set isTop if element is already at another position on the page

Reply:
state functions are used to guarantee current state value being up to date, they have nothing to do with the state of the dom